torch/optim/optimizer.py

```diff
@@ -167,11 +244,19 @@ class Optimizer:
             options (used when a parameter group doesn't specify them).
     """
 
-    def __init__(self, params, defaults):
+    def __init__(
+        self,
+        params: Union[Iterable[torch.Tensor], Iterable[Dict[str, Any]]],
+        defaults: Dict[str, Any],
+    ) -> None:
         torch._C._log_api_usage_once("python.optimizer")
         self.defaults = defaults
-        self._optimizer_step_pre_hooks: Dict[int, Callable] = OrderedDict()
-        self._optimizer_step_post_hooks: Dict[int, Callable] = OrderedDict()
+        self._optimizer_step_pre_hooks: Dict[  # type: ignore[valid-type]
+            int, Callable[[Self, Tuple[Any, ...], Dict[str, Any]], Optional[Tuple[Any, Dict[str, Any]]]]
```

Comment:
Bc you will not get an error in the following example from the description:

```python
optimizer.register_step_post_hook(hook_fn_other_optimizer)  # Parameter 1: type "Adam" cannot be assigned to type "Adagrad"
```

`PostHook` and `PreHook` accept any `Optimizer`. If you define a hook that specifically hints one subclass and then register it on another, you should get an error.

Reply:
Oh, in this case, do you think it'd be useful to break up the PreHook and PostHook into HookArgs, HookKwargs, and HookReturnType? Otherwise, the wins here are not very indicative.

Reply:
Yes you're right I should definitely try to make aliases for these! Thanks!